### DISCUSSION TITLE
Beginner Qs about changing line,marker, and color settings.

### DISCUSSION THREAD

**Discussion by mgroth0** [GSI-CS-CO/chart-fx#420] [Q&A]
My goal here is to create a scatter plot with two series and for each of those series I will also add an additional series which is a guassian fit. The gaussian fit should be visible as a line with no markers.

As far as colors go, I'm completely fine with the default. However, my attempts to change the line and marker settings seem to have an odd effect on the colors, as you'll see below.

I use a dark .css for my javafx app, which explains why most colors are inverted in the images below. This should be unrelated to the issue.

To start, here's my initial chart. It uses two `DoubleDataSet`s two `DefaultNumericAxis`s and an `XYChart`. Most of what you are seeing here are default settings:
![Screen Shot 2021-05-17 at 12 02 39 PM](https://user-images.githubusercontent.com/12831489/118520136-c902f380-b707-11eb-85ba-cb6df3a79eea.png)

In order to change line and marker settings, my search has lead me to beleive that `ErrorDataSetRenderer` is the proper way. But please correct me if I'm wrong. Applying the (kotlin) code below: 
```kotlin
chart.renderers.add(ErrorDataSetRenderer().apply {
		datasets.add(ds)
	  })
```

changes the plot to look like:
![Screen Shot 2021-05-17 at 12 07 08 PM](https://user-images.githubusercontent.com/12831489/118520855-69591800-b708-11eb-95cd-ebb8f72d6958.png)

Here you can see the color issue I mentioned above. The default colors for the lines have changed to shades of grey, and also the legend is now displaying the wrong colors.
 
Next I tried changing from lines to markers with custom colors. I tried:
```kotlin
	  ds.setStyle("markerColor=green; markerType=circle2;")
	  chart.renderers.add(ErrorDataSetRenderer().apply {
		markerSize = 10.0
		isDrawMarker = true
		polyLineStyle = LineStyle.NONE
		datasets.add(ds)
	  })
```

Expectation: a chart with no lines, green markers, and a legend that matches the markers.
Actual result: a chart with visible lines, green markers, and a legend that matches the lins instead of the markers.

![Screen Shot 2021-05-17 at 12 15 30 PM](https://user-images.githubusercontent.com/12831489/118522020-95c16400-b709-11eb-962d-0c7a84cfa97e.png)

I also tried adding `lineColor=transparent` to the style to make the lines go away, but this didn't work.

So my questions for now are:
1. How do you make the legend match the markers instead of the line?
2. How do you properly change the line color?
3. How do you properly make the line not show?

**Comment by RalphSteinhagen** [comment]
Hey,

> 2. How do you properly change the line color?

The line colour is actually set via 'strokeColor' (not 'lineColor') following the default JavaFX attribute styles. 
The style of individual DataSets are primarily set via 
  * `DataSet::setStyle("strokeColor=red; strokeWidth=3; markerType=circle;")` applicable to all data points and/or
  * `DataSet::.addDataStyle(index, "...")` for individual data points of a given DataSet.

You can find a list of globally known definitions in [XYChartCss](https://github.com/GSI-CS-CO/chart-fx/blob/89841136c67e0330bd5e7f1883bdf66469ea6dc1/chartfx-chart/src/main/java/de/gsi/chart/XYChartCss.java) though certain renderers may locally define additional ones. The various definitions for the default DataSet colours are defined in the [DefaultRenderColorScheme](https://github.com/GSI-CS-CO/chart-fx/blob/cd202c7c9a0b40f1e66b6d0fe8fd93fffbec5435/chartfx-chart/src/main/java/de/gsi/chart/renderer/spi/utils/DefaultRenderColorScheme.java) and can be further customised to your liking. The [LabelledMarkerSample](https://github.com/GSI-CS-CO/chart-fx/blob/89841136c67e0330bd5e7f1883bdf66469ea6dc1/chartfx-samples/src/main/java/de/gsi/chart/samples/LabelledMarkerSample.java) illustrates the various line and marker styles.

> 3. How do you properly make the line not show?

`errorRenderer.setPolyLineStyle(LineStyle.NONE);` should work. Have a look, for exmple, at the [ErrorDataSetRendererStylingSample](https://github.com/GSI-CS-CO/chart-fx/blob/89841136c67e0330bd5e7f1883bdf66469ea6dc1/chartfx-samples/src/main/java/de/gsi/chart/samples/ErrorDataSetRendererStylingSample.java).

> 1. How do you make the legend match the markers instead of the line?

The legend should follow the above definitions. You may need to 'nudge' and request to repaint the chart in case you have dynamically changed the style (for quick test resizing should also work).  The interfaces in `ErrorDataSetRenderer` are mostly for turning certain options on/off for all DataSets that are rendered by the Renderer (e.g. turning on/off of: error surfaces, error bars, bars, lines, marker, ...).
You may check whether you are actually modifying the Renderer you have attached the DataSets too ... this is a common mistake during development that a new renderer is added but the DataSets added to the old one and or not having added the renderer to the list of Renderer in the chart.

There are some additional samples [here](https://github.com/GSI-CS-CO/chart-fx/tree/master/chartfx-samples/src/main/java/de/gsi) that may be useful to explore the various options.

Hope this helps...  if you still are having some problem please feel free to copy-paste a short MVP example and we can go through this together.

<sub>Out of curiosity: may we ask the domain you are using our little lib?</sub>

**Reply by mgroth0** [reply]
Thanks for the immense amount of help again @RalphSteinhagen 

It turns there were couple things I needed to do to solve most of my problems:
- Use colons in the CSS instead of equals signs (e.g. `strokeColor: yellow` instead of `strokeColor=yellow`
- Use `Axis::forceRedraw` and `Legend::updateLegend` in the right places, after settings have been changed.

Now I've come upon some new issues with the legend:
- The grid behind the line is not always the right color (assuming its supposed to be the same as the line/marker color)
- The legend colors seem to only reflect the `strokeColor` property and not the `markerColor` property. However, `setPolyLineStyle(LineStyle.NONE)` does not remove the line and setting `strokeWidth:0` does not remove the line either. The only way I've been able to remove lines is with `strokeColor:transparent`, however this unintentionally removes the colors in the legend which I expect to match `markerColor`. 
- Also, I'd like the legend item shapes to match the marker shapes and the option to remove the grid behind the legend item graphics.

I've done a quick search through all the links you provided. In each I did a find for "legend", and found no examples of these kinds of settings being changed for the legend.

Here's an image demonstrating a few of these issues:
- Should the red grid behind the yellow line for 45 not be a yellow grid?
- Can I remove these grids altogether?
- Can I switch the lines in the legend into circles to match the markers?
![Screen Shot 2021-05-17 at 4 14 41 PM](https://user-images.githubusercontent.com/12831489/118550699-ff059f00-b72a-11eb-8227-9be5283c0d73.png)

In the above example I definitely set LineStyle.NONE to the correct renderers and also `strokeWidth: 0`. However the line still shows. If I leave everything the same but change strokeColor to transparent I get (notice legend is now incomplete):

![Screen Shot 2021-05-17 at 4 18 04 PM](https://user-images.githubusercontent.com/12831489/118551054-76d3c980-b72b-11eb-9d89-172fae956a10.png)

MVP:
```kotlin
package matt.mvp

import javafx.application.Application
import javafx.scene.Scene
import javafx.stage.Stage
import de.gsi.chart.XYChart
import de.gsi.chart.axes.spi.DefaultNumericAxis
import de.gsi.chart.renderer.LineStyle
import de.gsi.chart.renderer.spi.ErrorDataSetRenderer
import de.gsi.dataset.spi.DoubleDataSet
import javafx.application.Platform.runLater

fun main() = Application.launch(MVPExample::class.java)

const val LINE = false

class MVPExample: Application() {
  override fun start(primaryStage: Stage?) {
	val xAxis = DefaultNumericAxis()
	val yAxis = DefaultNumericAxis()
	val ds = DoubleDataSet(
	  "new dataset",
	  doubleArrayOf(1.0, 2.0, 3.0),
	  doubleArrayOf(1.0, 2.0, 3.0),
	  3,
	  false
	).apply {
	  /*setting strokeWidth to 0 doesn't make line go away*/
	  style = if (LINE) {
		/*legend grid-thing behind each legend item does not reflect strokeColor*/
		"strokeColor: yellow; strokeWidth: 0; markerColor: yellow; markerType: circle;"
	  } else {
		/*setting strokeColor to transparent makes legend have no indicator for the series, even with a markerColor set*/
		"strokeColor: transparent; strokeWidth: 0; markerColor: yellow; markerType: circle;"
	  }
	}
	val chart = XYChart(xAxis, yAxis).apply {
	  datasets += ds
	  renderers += ErrorDataSetRenderer().apply {
		datasets += ds
		markerSize = 10.0
		polyLineStyle = LineStyle.NONE /*has no effect*/
	  }
	}
	primaryStage!!.apply {
	  scene = Scene(chart)
	}.show()

	/*update and refresh everything at the very end*/
	runLater {
	  chart.axes.forEach { a -> a.forceRedraw() }
	  chart.legend.updateLegend(chart.datasets, chart.renderers, true)
	}
  }
}
```

**Reply by wirew0rm** [reply]
Hey,

the "grid" behind the line is actually the dashed error surface of the plot. If you do not manually override the error surface color, it will be assigned based on the default color scheme and the dataset index (by default the number of previously rendered datasets in the chart). You can manually override the error surface color from css or you can disable error surfaces by setting the renderer's error style to NONE.

hope this helps, alex

**Comment by dirri** [comment]
Trying for hours just to set the markerColor for the circle2 marker to red, but it always defaults somehow to blue.
I added a custom renderer and overwrite the defaultMarker with no luck. I´m gettin tired!
The funny thing is when i use the circle and not circle2 as marker it will be red. But i want to have circle2 painted in red! :-(

**Reply by wirew0rm** [reply]
see: https://github.com/fair-acc/chart-fx/issues/517
This will be fixed in the next release until then you will have to set the line color to control the marker outline. If you need the line in a different color you'll have to use different renderers or override the markers to also set the line color to fill color (don't forget to restore the graphics context afterwards).
